cardano-cli/src/Cardano/CLI/EraBased/Run.hs

```diff
@@ -22,72 +21,85 @@ import           Cardano.CLI.EraBased.Vote
 
 import           Control.Monad.Trans.Except
 import           Control.Monad.Trans.Except.Extra (firstExceptT)
-import           Data.Text (Text)
-
-newtype AnyEraCmdError
-  = AnyEraCmdGenericError ()
-
-renderAnyEraCmdError :: AnyEraCmdError -> Text
-renderAnyEraCmdError = \case
-  AnyEraCmdGenericError () -> "Generic any era command error"
+import           Data.Function ((&))
+
+data AnyEraCmdError
+  = AnyEraCmdGovernanceCmdError        !GovernanceCmdError
+  | AnyEraCmdEraDelegationError        !EraBasedDelegationError
+  | AnyEraCmdEraBasedRegistrationError !EraBasedRegistrationError
+  | AnyEraCmdEraBasedVoteError         !EraBasedVoteError
+  | AnyEraCmdGovernanceCommitteeError  !GovernanceCommitteeError
+  | AnyEraCmdGovernanceActionError     !GovernanceActionsError
+  deriving Show
+
+instance Error AnyEraCmdError where
+  displayError = \case
+    AnyEraCmdGovernanceCmdError e -> displayError e
+    AnyEraCmdEraDelegationError e -> displayError e
+    AnyEraCmdEraBasedRegistrationError e -> displayError e
+    AnyEraCmdEraBasedVoteError e -> displayError e
+    AnyEraCmdGovernanceCommitteeError e -> displayError e
+    AnyEraCmdGovernanceActionError e -> displayError e
 
 runAnyEraCommand :: ()
   => AnyEraCommand
   -> ExceptT AnyEraCmdError IO ()
 runAnyEraCommand = \case
   AnyEraCommandOf sbe cmd ->
-    firstExceptT AnyEraCmdGenericError $ shelleyBasedEraConstraints sbe $ runEraBasedCommand cmd
+    shelleyBasedEraConstraints sbe $ runEraBasedCommand cmd
 
 runEraBasedCommand :: ()
-  => EraBasedCommand era -> ExceptT () IO ()
+  => EraBasedCommand era -> ExceptT AnyEraCmdError IO ()
 runEraBasedCommand = \case
   EraBasedGovernanceCmds cmd -> runEraBasedGovernanceCmds cmd
 
 runEraBasedGovernanceCmds :: ()
   => EraBasedGovernanceCmds era
-  -> ExceptT () IO ()
+  -> ExceptT AnyEraCmdError IO ()
```

Comment:
Are the error types going to follow the command structure or are we doing one large sum type for all our errors?

Reply:
I reckon we should call this type `CmdError` and put it in `Cardano.CLI.EraBased.Error`.

Reply:
We already have `ClientCommandErrors` in Cardano.CLI.Run. Should we reuse this one as a top-level umbrella error type?

Reply:
Not yet.  It includes lots of legacy stuff as members.